mono/metadata/timer-ms.c

```diff
@@ -0,0 +1,332 @@
+		if ((thread->state & (ThreadState_StopRequested | ThreadState_SuspendRequested)) != 0)
+			mono_thread_interruption_checkpoint ();
+
+		mono_mutex_lock (&timer->domains_lock);
+
+		next_due_ticks = G_MAXUINT32;
+		for (i = 0, len = timer->domains->len; i < len; ++i) {
+			TimerDomain *tdomain = g_ptr_array_index (timer->domains, i);
+			if (tdomain && tdomain->next_due_ticks >= tdomain->last_fire_ticks)
+				next_due_ticks = MIN (next_due_ticks, tdomain->next_due_ticks);
+		}
+
+		mono_mutex_unlock (&timer->domains_lock);
+
+		if (mono_msec_ticks () < next_due_ticks) {
+			mono_gc_set_skip_thread (TRUE);
+			MONO_SEM_TIMEDWAIT_ALERTABLE (&timer->wakeup_sem, next_due_ticks - mono_msec_ticks (), TRUE);
+			mono_gc_set_skip_thread (FALSE);
+
+			if (mono_msec_ticks () < next_due_ticks) {
+				/* we might have been woken up by timer_wakeup */
+				continue;
+			}
+		}
+
+		mono_mutex_lock (&timer->domains_lock);
+
+		for (i = 0; i < timer->domains->len; ++i) {
+			TimerDomain *tdomain = g_ptr_array_index (timer->domains, i);
+
+			if (!tdomain)
+				continue;
+
+			if (mono_domain_is_unloading (tdomain->domain))
+				continue;
+
+			if (mono_msec_ticks () < tdomain->next_due_ticks)
+				continue;
+
+			tdomain->last_fire_ticks = mono_msec_ticks ();
+
+			mono_mutex_unlock (&timer->domains_lock);
+
+			mono_thread_push_appdomain_ref (tdomain->domain);
+			if (mono_domain_set (tdomain->domain, FALSE)) {
+				MonoObject *exc = NULL;
+				mono_runtime_invoke (app_domain_timer_callback_method, NULL, NULL, &exc);
+				if (exc)
+					mono_internal_thread_unhandled_exception (exc);
+
+				mono_thread_clr_state (thread , ~ThreadState_Background);
+				if (!mono_thread_test_state (thread , ThreadState_Background))
+					ves_icall_System_Threading_Thread_SetState (thread, ThreadState_Background);
+
+				mono_domain_set (mono_get_root_domain (), TRUE);
+			}
+			mono_thread_pop_appdomain_ref ();
+
+			mono_mutex_lock (&timer->domains_lock);
+		}
+
+		mono_mutex_unlock (&timer->domains_lock);
+	}
+}
+
+static void
+timer_ensure_running (void)
+{
+	g_assert (timer);
+
+	for (;;) {
+		switch (timer_status) {
+		case TIMER_STATUS_REQUESTED:
+			return;
+		case TIMER_STATUS_WAITING_FOR_REQUEST:
+			InterlockedCompareExchange (&timer_status, TIMER_STATUS_REQUESTED, TIMER_STATUS_WAITING_FOR_REQUEST);
+			break;
+		case TIMER_STATUS_NOT_RUNNING:
+			if (mono_runtime_is_shutting_down ())
+				return;
+			if (InterlockedCompareExchange (&timer_status, TIMER_STATUS_REQUESTED, TIMER_STATUS_NOT_RUNNING) == TIMER_STATUS_NOT_RUNNING) {
+				if (!(timer->thread = mono_thread_create_internal (mono_get_root_domain (), timer_thread, NULL, FALSE, 0)))
+					timer_status = TIMER_STATUS_NOT_RUNNING;
+				return;
+			}
+			break;
+		default:
+			g_assert_not_reached ();
+		}
+	};
+}
+
+static void
+timer_wakeup (void)
+{
+	g_assert (timer);
+
+	MONO_SEM_POST (&timer->wakeup_sem);
+}
+
+static guint32
+compute_next_due_ticks (guint32 due_time)
+{
+	guint32 now = mono_msec_ticks ();
```

Comment:
Time should not be tracked using guint32.

There are 86.400.000 ms in a day. Or 49.7 days worth of milliseconds in a guint32.
After that much of uptime (not process time) it will overflow and everything will behave weirdly.

Use mono_100ns_ticks instead.